man/dbshell.1
.TH DBSHELL 1 "21 September 1995"
.FB
.SH NAME
Dbshell \- the \fIFirstBase\fP menu tool.
.SH SYNOPSIS
.B dbshell
[\fB-d\fP \fIdatabase\fP] [\fB-i\fP \fIindex\fP]
[\fB-s\fP \fIscreen\fP]
[\fB-v\fP \fIview\fP]
[\fImenu\fP]
.SH DESCRIPTION
.I Dbshell
is the database menu shell tool which provides all of the flexibility
of the Bourne shell along with the power and ease of use that menus create.
Like other \fIFirstBase\fP tools, \fIdbshell\fP(1) works from a
menu dictionary \(em sort of.
.PP
The dbshell dictionary is more of a free form
text file created with your favorite editor (like \fIvi\fP(1)).
This file will be 'executed' by \fIdbshell\fP(1), allowing a smooth
interface between menu selections and the Bourne shell.
.PP
Using this dictionary,
\fIdbshell\fP(1) paints the screen and then provides a prompt for user input.
At this prompt, the user
issues commands. The command must be known by \fIdbshell\fP(1) or have been
built into the dictionary being executed.
\fIDbshell\fP(1) has a few built in commands. These are summarized in the
COMMANDS section below.
.PP
The default menu for \fIdbshell\fP(1) is \fImenu/MAIN\fP located in the
\fIfirstbase\fP home directory.
.PP
To execute a command, the entire selection does not need to typed \(em
just enough to distinguish it from other commands.
This feature can be disabled using the environment variable
MENUMATCH. See \fIsetup\fP(5) for more.
.PP
There are two major divisions to the dictionary/text file:
the \fIscreen\fP section and the \fIactions\fP section.
The \fIscreen\fP section of the dictionary file is all of the beginning
lines of the file, up to a line with a single percent sign.
A maximum of 21 lines are allowed.
.PP
This screen section will appear on the terminal \fIexactly\fP as depicted
in the file with one exception. To display
words or tokens in reverse video prepend a dollar sign (\fB$\fP) to the token.
A \fIdbshell\fP screen will have the
standard \fIFirstBase\fP header and footer lines, and
an "Enter Selection" prompt on the 24th line.
.PP
A token is defined as a series of
alphanumeric characters plus the underscore, or as punctuation. This means
characters like ampersand and comma must be marked individually.
Use an underscore to force reverse video of a blank.
For instance, in the example presented at the end of this section, the
areas 'Major Functions' and 'Minor Functions' will be displayed in reverse
video, as well as the '1' and the '2' (but not the periods).
.PP
The actions part of the dictionary/text file used with dbshell(1) consists
of simple shell commands or scripts
that are executed when their labels are selected. These commands
are located after the '%' marker in the file.
.PP
A label consists of a series of alphanumeric characters, with a dollar 
sign ('$') prepended to the label. In the example file presented, there are
five labels: $1, $2, $mail, $ls and $print. Each of the associated actions
connected to these labels will be executed as a shell script by the Bourne
shell.
.PP
So, an entry of '1' (and a <RETURN>) at the "Enter Selection: "
prompt would echo 'choice one' to the terminal. A '2' would change to
another directory and run the database editor. Etc.
.PP
There are three special flags that can be placed after a label. A 
negative one (-1),
as used in the '$ls' label, will cause dbshell(1) to NOT pause after returning
to the menu before the screen is redrawn.
.PP
The '*' flag is used to denote a sub menu dictionary/text file.
In this case, the sub menu is
selected by issuing the command 'print' (since it is tied to the $print
label). In the example, the actual menu file is \fIPRINT\fP.
Note that a full path name to \fIPRINT\fP is not needed as long as it
exists in the same directory as the calling menu. Thus,
systems of menus can be written using relative pathnames. Also, the menus can
be kept in one directory and invoked from another.
.PP
In this case, the initial command would be something like
.ft CW
dbshell menu_directory/MAIN.
.ft
This MAIN menu can then invoke the PRINT submenu as in the example below.
Again, \fIdbshell\fP will remember the \fImenu_directory\fP
of the first invoked menu. This directory name will be prepended to all
submenu names that are NOT full pathnames, i.e. that do not begin with a slash.
.PP
Optionally, after the '*', a directory name can be supplied. Using this
feature, \fIdbshell\fP(1) will change directories to the supplied place,
and then invoke the submenu. See the \fIwork\fP submenu choice below.
.PP
The optional working directory can also be a true UNIX environment
variable, prepended with a \fB$\fP like standard shell variable.
For example, if \fB$CTSHOME\fP was used, \fIdbshell\fP(1) will
substitute in the contents of the UNIX environment variable \fICTSHOME\fP.
.PP
The third special flag that can be placed after the label is the
meta command \fB$EXIT\fP. Furthermore, this line can contain a number
that will be used to exit, i.e. \fB$EXIT 31\fP would use the exit 
value \fI31\fP upon termination.
.PP
This special meta command will cause a forced exit from \fIdbshell\fP(1).
Any attached normal menu commands are executed before the exit takes place.
.SH COMMANDS
\fIDbshell\fP(1) has a few built in commands that are used from the command
line \(em the 'Enter Selection' prompt. These commands are summarized below.
.sp 1
.TP 10
.B cd
change directory. The user is asked for the name of the directory to
change to. If successful, the new directory name is displayed on the 
footer line. Otherwise an error message occurs.
.TP 10
.B <CTL>-E
This command displays the list of environment commands that can be used
to change various parts of the working environment. For example, you can
enter the names of the database and index, change working
directories, or even start a UNIX shell. See \fIChapter 3\fP of the
.I FirstBase User's Guide and Reference Manual
for a display of this screen. (Note that a <CTL>-D will
also work in place of <CTL>-E).
.TP 10
.B <CTL>-H
produce help screen. The help screens here are completely user defined.
\fIDbshell\fP(1) looks for a file with the same name as the menu, with an
extension of '.help', and pages this file to the screen. Help files can
be as long as needed.
.TP 10
.B -
Standard END keystroke. Ends from the current dictionary/menu. If this was
a sub menu, its parent is re-invoked. If this was the main menu (root
menu), the user will be asked 'Really Exit?'.
.TP 10
.B <CTL>-X
Abort from current menu.
The user will be asked 'Really Exit?'.
.SH EXAMPLE
Here is a very small sample dictionary/text file that could be used with
dbshell(1):
.nf
.nj
.na

      $Major_Functions       $Minor_Functions
      
       $1. Choice One           $mail - Send system mail
       $2. Edit Database        $ls   - list directory
       
                      $Sub_Menus
              $print - Printout Sub Menu
              $work  - Work Menu
              $quit  - Quit
%
$1
echo choice one

$2
cd /usr/databases/work
dbedit -d master -i coname -s invoice

$mail
echo -n "Send mail to who? "
read MAILTO
echo "Enter mail message now. End with a '.'."
mail $MAILTO

$ls -1
ls -F

$print *
PRINT

$work * /usr/work
WORK

$q $EXIT 0
echo Good Bye
.fi
.ju
.ad
.PP
Security Note: The standard locking method used to disable \fIdbedit\fP(1)
visual, filter and shell commands is implemented in \fIdbshell\fP(1) as
well. This disabling lock, \fI.cshell_lck\fP, must appear in either the
current directory or the system directory, \fI/usr/lib/firstbase\fP.
.PP
\fIDbshell\fP(1)
supports special meta-tokens that can be embedded within shell
actions. These tokens are listed below.
.sp 1
.TP 10
.B $FIRSTBASEHOME
appearing anywhere in the action
will be substituted at run time for the home directory of the firstbase system,
as defined by the UNIX environment variable of the same name.
.sp 1
This feature enables referencing of the \fIFirstBase\fP menu system from
within other application menu systems using 
.ft CW
$FIRSTBASEHOME/menu/MAIN
.ft
as the menu file referenced.
.TP 10
.B $DBASE
appearing anywhere in the action
will be substituted at run time for the argument line database.
.TP 10
.B $INDEX
appearing anywhere in the action
will be substituted at run time for the argument line index.
.TP 10
.B $SCREEN
appearing anywhere in the action
will be substituted at run time for the argument line screen.
.TP 10
.B $VIEW
appearing anywhere in the action
will be substituted at run time for the argument line view dictionary.
.TP 10
.B $RUNFLAGS
appearing anywhere in the action
will be substituted at run time for the environment variable string
defined for RUNFLAGS.
.PD
.SH FILES
.PD 0
.TP 10
main
name of default dbshell menu file.
.TP 10
/tmp/FBEXE??????
Temporary location of shell scripts. Controlled by TEMPDIR, via \fIsetup\fP(5)
.TP 10
*.help
extension for menu help files.
.TP 10
 .cshell_lck
sub process lock.
.PD
.SH SEE ALSO
dbshelltool(1), firstbase(1), setup(5).
.PP
.I FirstBase User's Guide and Reference Manual
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbshell/dbshell.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbshell.c,v 9.3 2001/02/12 23:25:05 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Dbshell_sid[] = "@(#) $Id: dbshell.c,v 9.3 2001/02/12 23:25:05 john Exp $";
#endif

/* 
 *  dbshell.c - another fb_database menu tool.
 *     provides a screen full of selections, with $words bound
 *     to series of commands, submitted as a shell process.
 */

#include <shell.h>
#include <shell_v.h>
#include <dbacct.h>
#include <setjmp.h>
#include <signal.h>

extern char *cdb_dbshell_pstring;
extern char *cdb_dbshell_ploc;
extern short int cdb_dbshell_pilength;
extern short int cdb_dbshell_ecommand;
extern short int cdb_locklevel;
extern short int cdb_secure;
extern short int cdb_menupause;
extern char *cdb_menu;
extern short int cdb_edit_input;
extern char *cdb_e_buf;
extern char *cdb_work_dir;
extern short int cdb_functionkeys;
extern short int cdb_dbshell_checkmail;	/* time between checkmail in dbshell */

short int force_exit = 0;
short int exit_value = 0;

char *PSTRING = "Enter Selection: ";
int ploc_row;
int ploc_col;
int screenmax;
static off_t mailsize;
static jmp_buf dbshell_jmp_env;

char menu_rootdir[FB_MAXNAME];

void set_mail();
void test_mail();

#if FB_PROTOTYPES
static void process(void);
static sh_select(void);
static command(char *w);
static rsh_usage(void);
static shell_redraw(void);
static RETSIGTYPE dbshell_sigalrm(int);
#else /* FB_PROTOTYPES */
static void process();
static sh_select();
static command();
static rsh_usage();
static shell_redraw();
static RETSIGTYPE dbshell_sigalrm();
#endif /* FB_PROTOTYPES */

/*
 *  dbshell - main driver 
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
      
      {
         FILE *fb_mustfopen();
	 int i, logshell = 0, rlogshell = 0;
	 char buf[FB_MAXLINE], *p, tname[FB_MAXNAME];
	 
	 cdb_locklevel = -1;
         /* call setup early to determine if cdb_secure */
         fb_setup_argv(argc, argv);
#if FB_CDB_SECURE
         if (cdb_secure){
            fb_basename(tname, argv[0]);
            if (equal(tname, "Fbsh"))
               logshell = 1;
            else if equal(tname, "fbrsh"){
               rlogshell = 1;
               if (fb_rlogin() == FB_ERROR)
                  fb_xerror(FB_MESSAGE, "Remote FirstBase Login failed.", NIL);
               if ((i = fb_testargs(argc, argv, "-x")) > 0){
                  i++;
                  if (i >= argc)
                     rsh_usage();
                  fb_system(argv[i], FB_NOROOT);
                  fb_settty(0);
                  fb_rlogout();
                  exit(0);
                  }
               }
            else
               logshell = 0;
            }
#endif /* FB_CDB_SECURE */
         fb_getargs(argc, argv, FB_ALLOCDB);
	 ploc_row = cdb_t_lines;
	 ploc_col = 1;
         screenmax = cdb_t_lines - 2;
         if (screenmax > SCREENMAX)
            screenmax = SCREENMAX;
	 if (cdb_dbshell_pstring == NULL)
	    fb_mkstr(&cdb_dbshell_pstring, PSTRING);
	 if (cdb_dbshell_ploc != NULL){
	    strcpy(buf, cdb_dbshell_ploc);
	    p = strchr(buf, ',');
	    *p = NULL;
	    ploc_row = atoi(buf);
	    ploc_col = atoi(p + 1);
	    }

         strcpy(cmenu, cdb_menu);
	 strcat(cmenu, "MAIN");
	 fb_rootname(dname, cdb_db->dbase);
	 fb_rootname(iname, cdb_db->dindex);
	 strcpy(sname, SCREENDEFAULT);
	 strcpy(vname, SCREENDEFAULT);
	 for (i = 0; i < SCREENMAX; i++)
	    screen[i] = NULL;
	 head->prev = tail->next = NULL;
	 head->next = tail;
	 tail->prev = head;
         for (i = 1; i < argc; i++)
            if (argv[i][0] == CHAR_MINUS){
               if (i  + 1 < argc){
	          if (argv[i][1] == CHAR_d)
		     fb_rootname(dname, argv[i + 1]);
		  else if (argv[i][1] == CHAR_i)
		     fb_rootname(iname, argv[i + 1]);
		  else if (argv[i][1] == CHAR_s)
		     fb_rootname(sname, argv[i + 1]);
		  else if (argv[i][1] == CHAR_v)
		     fb_rootname(vname, argv[i + 1]);
                  i++;
		  }
               }
            else if (strchr(argv[i], '=') == 0)
               strcpy(cmenu, argv[i]);
         fb_dirname(menu_rootdir, cmenu);
	 fs = fb_mustfopen(cmenu, "r");
	 if (fb_testargs(argc, argv, NON_CDB)){
	    cdb_db->dbase[0] = cdb_db->dindex[0] = NULL;
	    cdbmenu = 0;
	    }
	 else
	    fb_initdbd(cdb_db);
         fb_cx_set_menu(cmenu);
         fb_cx_push_env("EVH", CX_MENU_SELECT, NIL);
         set_mail();
         cdb_functionkeys = 0;
	 for (;;){
	    process();
	    if (force_exit)
	       break;
	    if (cdb_menupause == 0 ||
                  fb_mustbe(CHAR_y, EXITMSG, cdb_t_lines, 1) == FB_AOK)
	       break;
	    }
         fb_cx_pop_env();
	 fclose(fs);
#if FB_CDB_SECURE
         if (cdb_secure){
	    fb_acctlog(FB_AC_ENDTOOL, dname, iname);
            fb_clear(), fb_move(1,78);
            fb_printw(SYSMSG[S_FB]);
            fb_move(cdb_t_lines,1);
            fb_refresh();
            fb_settty(0);
            if (rlogshell)
               fb_rlogout();
            else if (logshell){
               execl("/fbetc/fblogout", "-fblogout", 0);
               printf("No FirstBase Logout Process.\n");
               }
            exit(0);
            }
#endif /* FB_CDB_SECURE */
	 if (force_exit){
	    fb_exit(exit_value);
	    }
	 fb_ender();
      }

/*
 * process loop of dbshell
 */

   static void process()

      {
         int st;
	 
	 initmenu();
         shell_redraw();
	 for (;;){
	    st = sh_select();
	    if (force_exit)
	       return;
	    if (st == FB_END)
	       break;
	    else if (st == FB_ERROR)
	       fb_serror(FB_MESSAGE, "Cannot Find That Command", NIL);
	    else{
	       if (cdbmenu)
	          fb_initdbd(cdb_db);
	       shell_redraw();
	       }
	    }
      }

/*
 * select - selection mechanism of dbshell
 */

   static sh_select()
      {
         char word[FB_MAXLINE], xdir[FB_MAXNAME], *fb_getwd();
	 int st, col;
	 struct an_rmenu *p;
         FILE *fb_mustfopen();
	 
	 for (st = 0;;){
            test_mail();
            /*
             * setjmp is either returning normally or from a signal (ALRM)
             */
	    setjmp(dbshell_jmp_env);
	    if (st != FB_DEFAULT){
	       if (ploc_row > 0){
		  fb_move(cdb_t_lines, 1); fb_clrtoeol();
		  fb_scrhlp(MSG2);
		  fb_move(ploc_row, ploc_col);
		  fb_printw(cdb_dbshell_pstring);
		  }
	       }
	    if (ploc_col > 0)
	       col = ploc_col + strlen(cdb_dbshell_pstring);
	    else
	       col = ploc_col;
            if (cdb_dbshell_checkmail > 0){
               signal(SIGALRM, dbshell_sigalrm);
               alarm(cdb_dbshell_checkmail);
               }
	    st = fb_input(-ploc_row, -col, -cdb_dbshell_pilength, 
	       0, FB_ALPHA, word, FB_ECHO, FB_OKEND, FB_CONFIRM);
            if (cdb_dbshell_checkmail > 0){
               signal(SIGALRM, SIG_IGN);
               alarm(0);
               }
	    if (st == FB_ABORT)
	       return(FB_END);
	    else if (st == FB_END){
	       if (head->next == tail)
	          return(FB_END);
	       strcpy(cmenu, tail->prev->rmenu);
	       strcpy(xdir, tail->prev->rdir);
	       fclose(fs);
	       chdir(xdir);
	       fb_mkstr(&cdb_work_dir, fb_getwd(xdir));
               fb_getco((fb_database *) NULL);
	       fs = fb_mustfopen(cmenu, "r");
               fb_dirname(menu_rootdir, cmenu);
	       p = tail->prev;
	       tail->prev = tail->prev->prev;
	       tail->prev->next = tail;
	       fb_free(p->rmenu);
	       fb_free((char *) p);
               fb_cx_set_workdir(xdir);
               fb_cx_set_menu(cmenu);
               fb_cx_write(1);
	       initmenu();
	       shell_redraw();
	       }
	    else if (st == FB_QHELP){
	       sprintf(word, "%s.help", cmenu);
	       fb_fhelp(word);
	       shell_redraw();
	       }
	    else if (cdb_dbshell_ecommand && 
	          (st == FB_DSIGNAL || st == FB_ESIGNAL || st == FB_YSIGNAL)){
               fb_cx_push_env("EH", CX_ENV_SELECT, NIL);
	       mcommand();
               fb_cx_pop_env();
	       return(FB_AOK);
	       }
	    else if (st == FB_AOK){
	       fb_rmlead(fb_trim(word));
	       if (command(word) == FB_AOK)
	          return(FB_AOK);
	       fseek(fs, seekpoint, 0);
               st = execute(word);
               return(st);
	       }
	    }
      }

/*
 * command()
 */

   static command(w)
      char *w;
   
      {
         char dir[FB_MAXLINE], *fb_getwd();
         int st;

         if (equal(w, "cd")){
            fb_cx_push_env("XT", CX_KEY_SELECT, NIL);
            fb_move(cdb_t_lines, 1); fb_clrtoeol();
            fb_printw("Change to What Directory? ");
            if (cdb_edit_input)
               strcpy(cdb_e_buf, fb_getwd(dir));
            st = fb_input(cdb_t_lines, 30, 50, 0, FB_ALPHA, dir, FB_ECHO,
               FB_OKEND, FB_CONFIRM);
            if (st == FB_AOK){
               fb_trim(dir);
               if (chdir(dir) != 0)
                  fb_serror(FB_MESSAGE, "Cannot change to directory:", dir);
               else{
                  fb_mkstr(&cdb_work_dir, fb_getwd(dir));
                  fb_initdbd(cdb_db);
                  fb_cx_set_workdir(dir);
                  fb_getco((fb_database *) NULL);
                  }
               }
            fb_cx_pop_env();
            return(FB_AOK);
            }
         else if (equal(w, "?b")){
            fb_help(w, cdb_db);
            shell_redraw();
            return(FB_AOK);
            }
         return(FB_ERROR);
      }

/*
 * shell_redraw - redraw the screen
 */

   static shell_redraw()
      {
         char token[FB_MAXLINE], *p;
	 int j, k, row;

	 fb_clear();
	 if ((p = strrchr(cmenu, '/')) == 0)
	    strcpy(pmenu, cmenu);
	 else
	    strcpy(pmenu, p+1);
	 fb_scrhdr(cdb_db, pmenu);
	 for (row = 2, k = 0; k < SCREENMAX; k++, row++){
	    p = screen[k];
	    if (p != NULL && *p){
	       fb_move(row, 1);
	       for (j = 1; (j = fb_gettoken(p, j, token, '_')) > 0; ){
		  if (token[0] != CHAR_DOLLAR)
		     fb_printw("%s", token);
		  else{
		     j = fb_gettoken(p, j, token, '_');
		     if (j <= 0)
			break;
		     fb_underscore(token, 0);
		     fb_reverse(token);
		     }
		  }
	       }
	    }
	 fb_scrtime(cdb_db);
      }

/*
 * initmenu - gather in the current screen into screen array.
 */

   initmenu()
      {
         char buffer[FB_MAXLINE];
	 int k = 0, j;
	 
	 for (j = 0; j < SCREENMAX; j++)
	    if (screen[j] != NULL){
	       fb_free(screen[j]);
	       screen[j] = NULL;
	       }
         fseek(fs, 0L, 0);
	 seekpoint = 0L;
	 while (fgets(buffer, FB_MAXLINE, fs) != NULL){
	    if (buffer[0] == CHAR_PERCENT){
	       seekpoint = ftell(fs);
	       break;
	       }
	    if (k >= SCREENMAX)
	       break;
	    fb_mkstr(&(screen[k++]), buffer);
	    }
	 if (seekpoint == 0L){
	    while (fgets(buffer, FB_MAXLINE, fs) != NULL){
	       if (buffer[0] == CHAR_PERCENT){
		  seekpoint = ftell(fs);
		  break;
		  }
	       }
	    }
	 if (seekpoint == 0L)
            fb_xerror(FB_MESSAGE, "dbshell MENU [database options] [-n]", NIL);
      }

/* 
 *  usage - usage for dbmenu 
 *
 * usage()
 *    {
 *       fb_xerror(FB_MESSAGE, "dbshell MENU [database options] [-n]", NIL);
 *    }
 */

/*
 * rsh_usage - usage for dbshell
 */

   static rsh_usage()
      {
         fb_xerror(FB_MESSAGE, 
            "fbrsh [-n] [-x command] [database options] menufile", NIL);
      }

/*
 * set_mail - init the email size. clear the email line.
 */

   void set_mail()
      {
         char mailfile[FB_MAXLINE], *p;
         struct stat ms;

         mailsize = 0;
         p = getenv("MAIL");
         if (p == NULL)
            return;
         strcpy(mailfile, p);
         if (stat(mailfile, &ms) == 0){
            mailsize = ms.st_size;
            }
      }

/*
 * test_mail -
 *    if no change in size, return
 *    size > old size, show You have new mail.
 *    size < old size, clear message
 */
 
   void test_mail()
      {
         char mailfile[FB_MAXLINE], *p;
         struct stat ms;
         off_t new_mailsize;

         p = getenv("MAIL");
         if (p == NULL)
            return;
         strcpy(mailfile, p);
         if (stat(mailfile, &ms) == 0)
            new_mailsize = ms.st_size;
         else
            new_mailsize = 0;
         if (mailsize < new_mailsize){
            fb_clear();
            fb_move(12,35);
            fb_reverse("You have new mail.");
            fb_refresh();
            cdb_e_buf[0] = NULL;
            FB_PAUSE();
            shell_redraw();
            if (stat(mailfile, &ms) == 0)
               new_mailsize = ms.st_size;
            else
               new_mailsize = 0;
            }
         mailsize = new_mailsize;
         fb_refresh();
      }

   static RETSIGTYPE dbshell_sigalrm(d)
      int d;

      {
         test_mail();
         fb_infoline();
	 longjmp(dbshell_jmp_env, 1);
      }
